src/core/proto/netlink_socket_mgr.cpp

```diff
@@ -4,173 +4,77 @@
-//      pid         : opaque pid for netlink request
-//      seq         : opaque seq for netlink request
-//      buf         : buffer for the request
-//      nl_msg      : [out] pointer to request
-void netlink_socket_mgr::build_request(nl_data_t data_type, uint32_t pid, uint32_t seq, char *buf,
-                                       struct nlmsghdr **nl_msg)
-{
-    struct rtmsg *rt_msg;
-
-    assert(MSG_BUFF_SIZE >= NLMSG_SPACE(sizeof(struct rtmsg)));
-    memset(buf, 0, NLMSG_SPACE(sizeof(struct rtmsg)));
-
-    // point the header and the msg structure pointers into the buffer
-    *nl_msg = (struct nlmsghdr *)buf;
-    rt_msg = (struct rtmsg *)NLMSG_DATA(*nl_msg);
-
-    // Fill in the nlmsg header
-    (*nl_msg)->nlmsg_len = NLMSG_LENGTH(sizeof(struct rtmsg));
-    (*nl_msg)->nlmsg_seq = seq;
-    (*nl_msg)->nlmsg_pid = pid;
-    (*nl_msg)->nlmsg_type = data_type == RULE_DATA_TYPE ? RTM_GETRULE : RTM_GETROUTE;
-    (*nl_msg)->nlmsg_flags = NLM_F_DUMP | NLM_F_REQUEST;
-
-    rt_msg->rtm_family = AF_UNSPEC;
-}
-
-// Query built request and receive requested data (Rule, Route)
-// Parameters:
-//      nl_msg  : request that is built previously.
-//      buf     : [out] buffer for the reply
-//      len     : [out] length of received data.
-bool netlink_socket_mgr::query(const struct nlmsghdr *nl_msg, char *buf, int &len)
+// Update data in a table
+void netlink_socket_mgr::update_tbl(nl_data_t data_type)
 {
-    int sockfd;
-
-    // Opaque information in the request. To track expected reply.
-    uint32_t nl_pid = nl_msg->nlmsg_pid;
-    uint32_t nl_seq = nl_msg->nlmsg_seq;
+    nl_sock *sockfd = nullptr;
 
     BULLSEYE_EXCLUDE_BLOCK_START
-    if ((sockfd = SYSCALL(socket, PF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE)) < 0) {
-        __log_err("NL socket creation failed, errno = %d", errno);
-        return false;
-    }
-    if (SYSCALL(fcntl, sockfd, F_SETFD, FD_CLOEXEC) != 0) {
-        __log_warn("Fail in fcntl, errno = %d", errno);
-    }
-    if ((len = SYSCALL(send, sockfd, nl_msg, nl_msg->nlmsg_len, 0)) < 0) {
-        __log_err("Write to NL socket failed, errno = %d", errno);
-    }
-    if (len > 0 && (len = recv_info(sockfd, nl_pid, nl_seq, buf)) < 0) {
-        __log_err("Read from NL socket failed...");
-    }
-    BULLSEYE_EXCLUDE_BLOCK_END
 
-    close(sockfd);
-    return len > 0;
-}
-
-// Receive requested data and save it to buffer.
-// Return length of received data.
-// Parameters:
-//      sockfd  : netlink socket
-//      pid     : expected opaque pid value
-//      seq     : expected opaque seq value
-//      buf     : [out] read reply
-int netlink_socket_mgr::recv_info(int sockfd, uint32_t pid, uint32_t seq, char *buf)
-{
-    struct nlmsghdr *nlHdr;
-    int readLen;
-    int msgLen = 0;
-    char *buf_ptr = buf;
+    sockfd = nl_socket_alloc();
+    if (sockfd == nullptr) {
+        __log_err("NL socket Creation: ");
+        throw_xlio_exception("Failed nl_socket_alloc");
```

Comment:
[question] Why not to return an error code instead of exception? To be compatible with the old API?